### ISSUE TITLE
[http-sync] Fileserver does not send response for large PUT requests (100 Continue related?)

### ISSUE THREAD

**Issue by klemens** [haiwen/seafile#1229]
Today I installed the new seafile version 4.2.0 and did some testing with the windows client 4.2.1.

There seems to be a problem with the fileserver not sending a http response for large put requests. The request itself it successfully received and saved as subsequent calls to `check-block` don't contain the block id anymore. However the client is waiting until it hits it's timeout: `libcurl failed to PUT *url*: Server returned nothing (no headers, no data).`

The complete seafile log from the client:
```
[05/14/15 15:19:56] sync-mgr.c(641): Repo 'test1' sync state transition from 'synchronized' to 'committing'.
[05/14/15 15:19:56] repo-mgr.c(2477): All events are processed for repo 7a7b20ac-3bfe-4bdc-b17f-9c107c3f01b7.
[05/14/15 15:19:56] sync-mgr.c(641): Repo 'test1' sync state transition from 'committing' to 'uploading'.
[05/14/15 15:19:56] http-tx-mgr.c(2390): Upload with HTTP sync protocol version 1.
[05/14/15 15:19:56] http-tx-mgr.c(674): Transfer repo '7a7b20ac': ('normal', 'init') --> ('normal', 'check')
[05/14/15 15:19:57] http-tx-mgr.c(674): Transfer repo '7a7b20ac': ('normal', 'check') --> ('normal', 'commit')
[05/14/15 15:19:58] http-tx-mgr.c(674): Transfer repo '7a7b20ac': ('normal', 'commit') --> ('normal', 'fs')
[05/14/15 15:19:58] http-tx-mgr.c(674): Transfer repo '7a7b20ac': ('normal', 'fs') --> ('normal', 'data')
[05/14/15 15:20:30] http-tx-mgr.c(524): libcurl failed to PUT http://seafile.domain.test/seafhttp/repo/7a7b20ac-3bfe-4bdc-b17f-9c107c3f01b7/block/a9597c65ccd8a01a3dd2614ab303ba433f6de743: Server returned nothing (no headers, no data).
[05/14/15 15:20:30] http-tx-mgr.c(2512): Failed to send block a9597c65ccd8a01a3dd2614ab303ba433f6de743 for repo 7a7b20ac.
[05/14/15 15:20:30] http-tx-mgr.c(674): Transfer repo '7a7b20ac': ('normal', 'data') --> ('error', 'finished')
[05/14/15 15:20:30] sync-mgr.c(708): Repo 'test1' sync state transition from uploading to 'error': 'Error occured in upload.'.
[05/14/15 15:20:30] sync-mgr.c(641): Repo 'test1' sync state transition from 'synchronized' to 'uploading'.
[05/14/15 15:20:30] http-tx-mgr.c(2390): Upload with HTTP sync protocol version 1.
[05/14/15 15:20:30] http-tx-mgr.c(674): Transfer repo '7a7b20ac': ('normal', 'init') --> ('normal', 'check')
[05/14/15 15:20:30] http-tx-mgr.c(674): Transfer repo '7a7b20ac': ('normal', 'check') --> ('normal', 'commit')
[05/14/15 15:20:31] http-tx-mgr.c(674): Transfer repo '7a7b20ac': ('normal', 'commit') --> ('normal', 'fs')
[05/14/15 15:20:31] http-tx-mgr.c(674): Transfer repo '7a7b20ac': ('normal', 'fs') --> ('normal', 'data')
[05/14/15 15:20:32] http-tx-mgr.c(674): Transfer repo '7a7b20ac': ('normal', 'data') --> ('normal', 'update-branch')
[05/14/15 15:20:32] http-tx-mgr.c(674): Transfer repo '7a7b20ac': ('normal', 'update-branch') --> ('finished', 'finished')
[05/14/15 15:20:32] sync-mgr.c(641): Repo 'test1' sync state transition from 'uploading' to 'initializing'.
[05/14/15 15:20:32] sync-mgr.c(1461): Removing blocks for repo test1(7a7b20ac).
```

The http conversation to the reverse proxy looks like this: (`HTTP/1.1 100 Continue` is really at the end because it is cached by the reverse proxy)
```
PUT /seafhttp/repo/7a7b20ac-3bfe-4bdc-b17f-9c107c3f01b7/block/a9597c65ccd8a01a3dd2614ab303ba433f6de743 HTTP/1.1
Host: seafile.domain.test
Accept: */*
User-Agent: Seafile/4.2.1 (Windows NT)
Seafile-Repo-Token: 6fe1618812522e555100995fac5119fa7614b468
Content-Length: 1672002
Expect: 100-continue

[data]
> HTTP/1.1 100 Continue
> 
```

The conversation between the proxy and the fileserver looks like this: (`HTTP/1.1 100 Continue` is not really at the end, but received after around 44k bytes have been transfered)
```
PUT /repo/7a7b20ac-3bfe-4bdc-b17f-9c107c3f01b7/block/a9597c65ccd8a01a3dd2614ab303ba433f6de743 HTTP/1.1
Connection: close
X-Hiawatha-RProxy-ID: 8rc80586f0
Host: seafile.domain.test
Accept: */*
User-Agent: Seafile/4.2.1 (Windows NT)
Seafile-Repo-Token: 6fe1618812522e555100995fac5119fa7614b468
Content-Length: 1672002
Expect: 100-continue
Forwarded: for="::ffff:81.28.133.22"
X-Forwarded-For: ::ffff:81.28.133.22
X-Forwarded-Proto: http
X-Forwarded-Host: seafile.domain.test
X-Forwarded-Port: 80

[data]
> HTTP/1.1 100 Continue
> 
```

As already noted, the reverse proxy ([hiawatha](https://github.com/hsleisink/hiawatha/)) does not really support 100 Continue currently, it immediately starts sending the request data without waiting for the `HTTP/1.1 100 Continue` and caches the `HTTP/1.1 100 Continue` on responses until all data is sent. This may confuse the fileserver.
Some more tests seem to show a correlation between "no answer" and "`100 Continue` send while transferring request". The turnaround point seems to be between about 450 and 490 KiB, with files 450 KiB or smaller working without problems (PUTs for commits are also working without problems).
```
PUT /seafhttp/repo/7a7b20ac-3bfe-4bdc-b17f-9c107c3f01b7/block/3fa4aadbc9386121363e86125090db0f681a7aea HTTP/1.1
Host: seafile.domain.test
Accept: */*
User-Agent: Seafile/4.2.1 (Windows NT)
Seafile-Repo-Token: 6fe1618812522e555100995fac5119fa7614b468
Content-Length: 22
Expect: 100-continue

[data]
> HTTP/1.1 100 Continue
> 
> HTTP/1.1 200 OK
> Content-Length: 0
> Connection: close
> 
```

**Comment by killing** [comment]
Can you use Nginx or Apache? We haven't tested with hiawatha.

2015-05-14 22:33 GMT+08:00 Klemens Schölhorn <notifications@github.com>:

**Comment by klemens** [comment]
I could do some tests with nginx if this helps you, but hiawatha is my primary web- and proxy-server. I´ve been using it for almost two yeas with seafile and there has never been a problem.
And as the second conversation between hiawatha and the fileserver (extracted from a tcpdump) shows, this seems to be a problem of the fileserver (or the underlying libevhtp) which is not sending the expected final respone.

But I can understand if this is a low priority issue for you and I will try to debug this further myself.

**Comment by StefanKrauth** [comment]
@klemens: Looks like the same issue lighttpd is producing. See [Issue 1230](https://github.com/haiwen/seafile/issues/1230) I created. Lighttpd has a switch to change the behavior and ignore the expected header. It works for the command line curl and for many other applications like wordpress swfuploader. However not for seafile.

Does hiawatha provide the ability to modify header data? So maybe you could modify the request and strip the Expected header?

Another way would be a fix on seafile side to strip out the Expected header from the put request using [libcurl curl_easy_setopt CURLOPT_HTTPHEADER](http://curl.haxx.se/libcurl/c/CURLOPT_HTTPHEADER.html).

This would probably be a good idea to create a more compatible request. See CURL [FAQ 4.16](http://curl.haxx.se/docs/faq.html)

**Comment by klemens** [comment]
Currently there is no possibility in hiawatha to change headers on proxy requests, but I could compile a version that strips the `Expect` header to see if this helps.

Generally though, disabling `Expect: 100-continue` in the client when it is not used in the server seems the way to go.

**Comment by killing** [comment]
Yes, I think we can remove the header in Seafile.

2015-05-16 20:20 GMT+08:00 Klemens Schölhorn <notifications@github.com>:

**Comment by StefanKrauth** [comment]
@killing that would be awesome!

**Comment by StefanKrauth** [comment]
@killing Do you have this on your todo list? No pressure, I just would like to know as I'm kind of waiting for this.

**Comment by mb3dot** [comment]
@killing +1 Same problem hier with Lighty and Seafile Pro 4.2.0. Can you estimate how long it takes to change this in Seafile?

**Comment by killing** [comment]
The latest client fixed this issue. Please have a try.

**Comment by StefanKrauth** [comment]
Hi @killing,
thank you very much! I'll try it on windows and mac as soon as the 2.4.5 client is released.

**Comment by StefanKrauth** [comment]
@killing:

I have just tested the windows version. I'm still running into the same error:

<pre>
http-tx-mgr.c(3474): Download with HTTP sync protocol version 1.
[06/24/15 15:18:15] http-tx-mgr.c(931): Transfer repo '7b76f2be': ('normal', 'init') --> ('normal', 'check')
[06/24/15 15:18:15] http-tx-mgr.c(931): Transfer repo '7b76f2be': ('normal', 'check') --> ('normal', 'commit')
[06/24/15 15:18:15] http-tx-mgr.c(931): Transfer repo '7b76f2be': ('normal', 'commit') --> ('normal', 'fs')
[06/24/15 15:18:15] http-tx-mgr.c(870): libcurl failed to POST https://*****/seafhttp/repo/83a1a5c6-f424-4247-91b3-d479a3740d28/pack-fs/: Failure when receiving data from the peer.
[06/24/15 15:18:15] http-tx-mgr.c(3512): Failed to get fs objects for repo 83a1a5c6 on server https://*****.
[06/24/15 15:18:15] http-tx-mgr.c(931): Transfer repo '83a1a5c6': ('normal', 'fs') --> ('error', 'finished')
[06/24/15 15:18:15] clone-mgr.c(812): Transition clone state for 83a1a5c6 from [fetch] to [fetch].
</pre>

I didn't check the webserver side yet. I can do this over the weekend earliest.

Best Regards,
Stefan

**Comment by klemens** [comment]
I have just checked with the new client version 4.2.6 and there is no `Expect`-header anymore. Therefore it works much better now:

Small to medium (a few hundred MiB) sized files work almost without problems, sometimes one or two blocks fail on the first try, but are then uploaded successfully on the second try. [Example with 160 MiB file](https://gist.github.com/klemens/b5b1abce739bb2b9cd39#file-medium-size-file-160-mib)

However this becomes a problem for larger files, because when one block-upload fails, the repo goes to the error state and waits 10-30 seconds before starting the next try. For large files this can result in many "upload-cycles": Uploading a 1.2 GiB file for example took almost 4 minutes because of the delays between retries: [Example with 1.2 GiB file](https://gist.github.com/klemens/b5b1abce739bb2b9cd39#file-large-file-1-2-gib)

@StefanKrauth:

I don't see your problem here, but I also don't have any `pack-fs` calls. (?)

**Comment by StefanKrauth** [comment]
I just figured out that I have one small library which syncs just fine. All other libraries are stopping with these `pack-fs` calls. What are these calls doing? I couldn't figure it out from the source code.

This is the webserver log: https://gist.github.com/StefanKrauth/19bbcdc2214edcb18578

**Comment by klemens** [comment]
It seems the `Expect`-header was only disabled for PUT requests but not for POST ones: e5cf8b2fc9368437be568d745c6458c429b746ea
I have tested with a fresh server install and therefore only a small library. Maybe this is why it is (almost) working for me. (as there is no need for `pack-fs` with a small and/or new library?)

**Comment by StefanKrauth** [comment]
Line 818 should handle the http_post events, no?

**Comment by klemens** [comment]
Yes, but the `Expect`-header is only disabled for PUT-requests: (see line [705](https://github.com/haiwen/seafile/blob/e5cf8b2fc9368437be568d745c6458c429b746ea/daemon/http-tx-mgr.c#L705))
```c+
headers = curl_slist_append (headers, "Expect:");
```

**Comment by killing** [comment]
Sorry I missed that. This will be fixed too ;-)

2015-06-30 2:31 GMT+08:00 Klemens Schölhorn <notifications@github.com>:

**Comment by StefanKrauth** [comment]
Hi, I've just tried the 4.2.7 release on windows - works great now! Also behind a HTTP proxy. Thanks for the fix @killing.

**Comment by klemens** [comment]
For the record, the `Expect`-header for PUT-requests has been disabled in a3e5e736459ac71a31045793bde4fd805b1d34c3.
I will close this issue, as it seems to be working for lighttpd and mostly for hiawatha now. Thank you very much!

If I find out the cause for some of the blocks failing (see https://github.com/haiwen/seafile/issues/1229#issuecomment-116285464), I will open a new issue.
